### PR TITLE
implement debugging support for dotnet

### LINKED ISSUE #333: Debugger Support: .NET
Add debugging support to the .NET language host.

### PR DESCRIPTION
**Pull request by tgummerer**
For dotnet to run with a debugger, we don't need to run the binary under a debugger, but rather the IDEs can just attach to the program. This needs some support from the SDK, so we wait until the debugger is actually attached.

Note that this does not seem to work when just running the program using `dotnet run .`, but rather we need to build and publish the program first, and only then we can attach the debugger to it.  (It might just be that running the program under the debugger is horribly slow, and thus using `dotnet run` takes more time than I had patience for, but that wouldn't be a good user experience either way).

I'm not sure how to write an integration test for this.  For Go the debugger starts a tcp server we can communicate with, but not sure what attaching to a PID means in that context/how exactly to do that.  I did try it out manually (which was also difficult, since including the local SDK doesn't seem to work with `dotnet build; dotnet <path-to.dll>`, so I ended up just copying the SDK changes to a program :grimacing:)

Fixes #333

### CODE REVIEW

File: sdk/Pulumi/Deployment/Deployment_Run.cs [hunk 1]
```diff
@@ -18,6 +18,15 @@ public partial class Deployment
         public static Task<int> RunAsync(Action action)
             => RunAsync(() =>
             {
+                var value = System.Environment.GetEnvironmentVariable("PULUMI_ATTACH_DEBUGGER");
+                if (value != null && value == "true")
+                {
+                    while (!System.Diagnostics.Debugger.IsAttached)
+                    {
+                        // keep waiting until the debugger is attached
+                        System.Threading.Tasks.Task.Delay(1);
```
Comment (Frassle) [inline sdk/Pulumi/Deployment/Deployment_Run.cs:27]: Need to await this

File: pulumi-language-dotnet/main.go [hunk 2]
```diff
@@ -547,8 +559,49 @@ func (w *logWriter) LogToUser(val string) (int, error) {
 	return len(val), nil
 }
 
+func (host *dotnetLanguageHost) buildDll() error {
+	// If we are running from source, we need to build the project.
+	// Run the `dotnet build` command.  Importantly, report the output of this to the user
+	// (ephemerally) as it is happening so they're aware of what's going on and can see the progress
+	// of things.
+	args := []string{"build", "-nologo"}
```
[the remark this answers is not in the document]
Comment (tgummerer) [reply pulumi-language-dotnet/main.go:567]: Hmm is that not the default profile?  At least when I just run `dotnet build`, the artifacts end up in a `bin/Debug/...` directory.

File: pulumi-language-dotnet/main.go [hunk 3]
```diff
@@ -547,8 +559,49 @@ func (w *logWriter) LogToUser(val string) (int, error) {
...
+
+	cmd := exec.Command(host.exec, args...)
+	err := cmd.Run()
+	if err != nil {
+		return errors.Wrapf(err, "failed to build project: %v", err)
+	}
+
+	args = []string{"publish", "-o", "bin/publish"}
```
[the remark this answers is not in the document]
Comment (tgummerer) [reply pulumi-language-dotnet/main.go]: I thought it was needed, but looks like just build is enough after all.  Removed the publish step.

File: sdk/Pulumi/Deployment/Deployment_Run.cs (hunk 1 above)
Comment (tgummerer) [reply sdk/Pulumi/Deployment/Deployment_Run.cs]: We're not in an `async` context here, so I switched it out for `System.Threading.Thread.Sleep(1);`.  Or is there a strong reason why we should be using `Delay`?

Comment (Frassle) [reply sdk/Pulumi/Deployment/Deployment_Run.cs]: Delay is better if in an async context. Sleep works fine here.

File: pulumi-language-dotnet/main.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (Frassle) [reply pulumi-language-dotnet/main.go]: It should be the default for most projects. Just gonna be one of those annoying edge case things that'll be grim to fix and we'll have to hope doesn't come up.

File: pulumi-language-dotnet/main.go [hunk 4]
```diff
@@ -547,8 +559,42 @@ func (w *logWriter) LogToUser(val string) (int, error) {
...
+		return errors.Wrapf(err, "failed to build project: %v", err)
+	}
+
+	err = filepath.WalkDir(".", func(path string, d os.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if name, ok := strings.CutSuffix(d.Name(), ".csproj"); ok {
+			host.binary = filepath.Join("bin", "pulumi-debugging", name+".dll")
```
Comment (Frassle) [inline pulumi-language-dotnet/main.go:581]: This probably shouldn't change global state. A language plugin can be used for multiple things in parallel and this is gonna confuse things.
Honestly "binary" shouldn't even be a global var to start with, the host should be re-parsing it from options on each run.

File: pulumi-language-dotnet/main.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (tgummerer) [reply pulumi-language-dotnet/main.go]: :crossed_fingers:

File: pulumi-language-dotnet/main.go (hunk 4 above)
Comment (tgummerer) [reply pulumi-language-dotnet/main.go]: Changed that now.  I haven't fixed the global `host.binary` though, that's probably something to look at in a separate PR.

**Pull request merged**